### ISSUE TITLE
Sync mode (client server) - strange things happen in protected sessions...

### ISSUE THREAD

**Issue by sharpeagle2022** [zadam/trilium#3111]
### Trilium Version

0.54.3

### What operating system are you using?

Windows

### What is your setup?

Local + server sync

### Operating System Version

W10 21H1

### Description

Please, explain a bit the concept of encryption and how the "protected sessions" and defined passwords (locally and on server) are working, because I am experiencing something very strange here....
I've read [this](https://github.com/zadam/trilium/wiki/Protected-notes#encryption-details) but the real life behavior seems different than the process described there....

So, I have a trilium.cc web instance and a local instance (started fresh with sync configured at 1st start). Everything fine until now...

But...
What I initially had in mind as a setup concept, thinking that this would work (it did not - it's clear for me by now why, eg there cannot be multiple/different passwords on client and server; but along the way I discovered something very strange, bear with me here till the end, please):
- to have 2 main parent notes under root (so that both can be encrypted, one using the locally set password, the other with the server set password)
- 1st one created to be created while the client is offline, lets call it PARENT1 and protected by local pwd
- 2nd one to be created on the server, PARENT2, protected by server pwd
- putting then the client back in sync with the server and assuming  that PARENT1 will be visible with local pwd only and PARENT2 with server password only.....
Sort of a multi-user mode on the same server. But just in my imagination....Cause I discovered during my experiments is that if a password is changed on one side, it is automatically pushed on the other side. Makes sense, probably the whole DB is synced. But, before finding this out....

What I did in order to attempt to have this setup in place:
1. I closed the client and disconnected the machine from the internet
2. reopened the client, set a local password (of course, different than the one set initially `1on the server)
3. created locally PARENT1, right click on it, "protect sub-tree", typed the password defined at point 2, entered "protected session", start creating some child notes, closing protected session
4. closed application, reopen it, typed the local passwords and the child notes are visible, everything fine
5. closed the application again and reconnect the machine to internet
6. opened again local application and pushed the sync button
7. back in the web interface on the server side I see that "protected note" appeared (transferred from local instance)
8. Now the tricky part.... I said to myself "let's try the server's password and see what is happening". Guess what: I was able to see the child note after typing the server's password in the web browser....
_Question here_: that note should have been visible only after being decrypted with the password defined locally, right? But still... server had opened it with it's own (different) password....

Tried the same scenario, but in a reversed order. Meaning, closed the client, created a note on the server, encrypted it and so on, as similar with above. Again, the same thing: client was able to see the cloud created and encrypted note by using its locally defined password....

As I said, what I noticed along the way is that at some point the passwords are silently synced, haven't noticed the exact context when this happens or after how long....
Still, what I don't understand is how point 8 above was even possible until password sync?
I tried multiple times in a row, opened protected session locally with local password, notes are visible, closed the local session, went back in server, opened the notes with server's password, notes are again visible (and event the content typed locally is visible on the server). And while playing around, passwords get synchronized (usually from local to server). Then, both require the same password to show the same notes (I believe it was the latest modified one in the DB, be it on server or on the client).

-----------------------

So....
How is this process taking place? The only possible explanation may be that:
1. prior to notes being sent on the wire , they are decrypted (well, they are still protected in a SSL tunnel), and when they arrive at the destination, are again encrypted but this time with the receiver's password? 
2. or....are the notes actually not encrypted at all (this process is somehow skipped) and it is just a matter of application deciding to display a "protected" note (the notes that have a property or something attached them in the database, while they in fact sit there unencrypted) only if the password entered by the user matches a password defined in the DB? And until the password sync takes place, both ends are able to display the synced notes with their respective passwords?

Sorry for the long post, but there was no other way to detail this...
Thanks!

**Comment by zadam** [comment]
Hi, the encryption is 2 phase:

1) there's a secret "data key" which is used to encrypt the actual data
2) "data key" is encrypted using the password

So if you change password, we just re-encrypt the "data key" (which itself stays the same) and don't have to re-encrypt the whole data set.

So if you change password on one side, there is a window during which you can still use the old password to decrypt the data, until the new password is synced to the other instance, because both encrypted strings lead to the same decrypted "data key".

In normal operation this should be instantaneous, I kinda assume that you marked as protected somewhat large subtree so the sync process took a while to finish.

**Comment by sharpeagle2022** [comment]
> Hi, the encryption is 2 phase:
> 
>     1. there's a secret "data key" which is used to encrypt the actual data
> So if you change password, we just re-encrypt the "data key" (which itself stays the same) so that we don't have to re-encrypt the whole data set.
To be honest, I thought of this scenario also, but did not include it in the "possible explanation" part, reason being: i assumed this would be like describing a security flaw in plain sight (see below why)....

It is now clear for me that "data key" is unique, randomly generated at DB creation (wiki says: "_data encryption key is random key generated at the time of [document](https://github.com/zadam/trilium/wiki/Document) initialization and is constant over the lifetime of the document. If we change password, we re-encrypt only this key_"
This explains why one cannot use multiple passwords to encrypt nodes separately. And also why there is a password needed at 1st install (just to generate and encrypt the "data key").

But, the question arises: isn't this concept a bit flawed in terms of security? Meaning, if the "data key" (I guess it's "encryptedDataKey" in the DB) cannot be actually changed (periodically), then if someone grabs the DB and does a brute force on the key itself, at some point that someone will be able to decrypt all notes, irrespective of the passwords the user is configuring over time (the user thinking that this is a good practice in preventing the DB hijacking...so, being deceived (s)he is on the safe side).
I guess it would be a good idea to allow the users to change also the "data key" (even if this means re-encrypting everything, so, a long process if the DB is big).

----------

Another aspect is this: I think it would be better if the server login password is different than the one used to decrypt the 
encryptedDataKey in the DB.  Just as a "multi-factor authentication" idea...And also to prevent locking the web interface if one is able to get the decrypted data key from the DB at some point (eg one cannot log in the interface with that key). Or to prevent accessing the encrypted notes if one is able to get the login password on the server....
This will also pave way for the potential to have the situation described below.

--------

On the other hand, do you thing it's doable to implement the encryption in a way where multiple keys can be defined? (eg my concept of multiple users on the same DB coupled with the inability to delete notes if they are encrypted). In this scenario, login is possible with one of the passwords and user will browse only the key-associated encrypted notes). If the user wants to browse a branch encrypted with another password, then the current encrypted session is closed and a new one is opened with that password.
I guess each note will then have a key ID associated, pointing to what key shall be used to decrypt it (provided the note is encrypted). So, a new table in DB with those keys and their IDs,
The drawback being that encrypted sessions are isolated, so no cloning can be made between trees encrypted with different keys (unless it is allowed to be cloned as an unecrypted note in another encrypted tree - user has to decide). Another issue may be with already included notes (referenced) in different trees, prior to protecting one of the trees: after encryption, the reference will be displayed in the other trees as encrypted string. Or not displayed at all, but instead a "Note encrypted" box. 
Or the inability to encrypt a note (or a tree that has notes) referenced in another place, or if the tree has already an encrypted sub-node/child note with a different key than the one in the active session etc.
I'm sure there are many other scenarios where adaptations are needed, so that the user experience is not broken at some point....

Thanks!

**Comment by zadam** [comment]
> then if someone grabs the DB and does a brute force on the key itself, at some point that someone will be able to decrypt all notes, irrespective of the passwords the user is configuring over time (the user thinking that this is a good practice in preventing the DB hijacking...so, being deceived (s)he is on the safe side).

If user chooses a strong password (which is of course recommended), then brute forcing it is more or less impossible (it goes through scrypt strengthening).

You're right that rotating the internal data key would slightly increase security parameters, but I don't think it will matter that much in practice and it will force reencryption of all notes (with many GB documents might be a problem).

> On the other hand, do you think it's doable to implement the encryption in a way where multiple keys can be defined? 

It is possible, but it's not desired. Trilium uses one password to simplify the UX. Trilium is a single user application, having more than one password does not seem necessary aside from niche use cases.

**Comment by sharpeagle2022** [comment]
In this case (while still being a single-user app), there could be an option to create multiple databases from within the GUI (located on different file system paths)  and the ability to select for each database the login is to be performed and its associated password.

Some of the users prefer to keep individual databases, each for a specific domain/purpose and distinct work-spaces. While Hoist-Note feature is good, it's just a simulation of a real isolated workspace (just a simple TAB). Also, size consideration, cluttering the database, having the graph more readable/relevant (less totally unrelated notes present there), performance etc.

The main drawback with using the "TRILIUM_DATA_DIR" method has to do with client-server mode coupled with VPS or various cloud providers. It is complicated to be managed and may imply in some cases a new subscription for each of the new desired database....When this could have been achieved in the same instance, if the above features were present.
Again, this is in the context of single-user usage....but can easily be extended (conceptually and functionally only, depending how you view it) to a multi user kind of usage (but probably not simultaneous usage of databases).

**Comment by zadam** [comment]
> Some of the users prefer to keep individual databases, each for a specific domain/purpose and distinct work-spaces. 

The "canonical" solution to this is hoisting/workspaces. One thing which I learned during the development of Trilium is that providing multiple ways to achieve essentially the same thing is a bad idea. Yes, it provides choice to user, but at the cost of often much increased complexity for a little benefit. It's usually better investment of time to try to improve the existing "one way" or implementing a feature which doesn't exist at all.

**Comment by sharpeagle2022** [comment]
> One thing which I learned during the development of Trilium is that providing multiple ways to achieve essentially the same thing is a bad idea. 

Yes, I'm with you on this one. And I perfectly understand that you are doing this in your spare time, simply because you love the this piece of software and love to code :)
So, my purpose was to convince you about an idea so that you end-up liking the idea and implementing it. 

Cause what I proposed above does not imply "providing the same ways to achieve essentially the same thing" ("TRILIUM_DATA_DIR" is not the same thing and not intrinsic to Trilium) or fundamental changes to the core (or too much to  code, I guess...). It was about a new feature, an UI being displayed prior to the engine launch that may allow the creation of new database or opening an existing database (if no active session present on another one). Once the DB is selected, the product does what is currently doing. I think the work here has to be done rather on the middlewares than in the core code itself (ie docker, node, restarting the services etc).
But that's ok, if you don't think this will appeal to many users, I get it and I appreciate your work nevertheless.
Thanks anyway!

**Comment by eliandoran** [comment]
Closing because the questions were answered and regarding the potential use of multiple databases, there are other feature requests for it.
